hypervisor/src/arch/x86/mod.rs

```diff
@@ -306,16 +310,191 @@ pub struct MsrEntry {
+        let actual_fam_length = self.0.as_fam_struct_ref().len;
+        let bytes_per_entry = size_of::<kvm_bindings::__u32>();
+        if *required_fam_length != actual_fam_length {
+            return Err(
+                format!("The required xsave struct size has changed since the initialization of this instance. The absolute difference in bytes is: {}",
+                (*required_fam_length * bytes_per_entry)
+                    .abs_diff(actual_fam_length * bytes_per_entry)
+            ).into());
+        }
+
+        store_routine(&self.0).map_err(Into::into)
+    }
+
+    pub fn enable_amx_state_components(
+        hypervisor: &dyn Hypervisor,
+    ) -> Result<(), AmxGuestSupportError> {
+        // It is only necessary to enable AMX related state state components for guests
+        // once, but this method may be called multiple times in the presence of live-migration
+        // aborts, or in presence of the upcoming CPU profiles feature.
+        let mut amx_enabled_guard = AMX_ENABLED_FOR_GUESTS.write().unwrap();
+        if *amx_enabled_guard {
+            return Ok(());
+        }
+
+        Self::amx_supported(hypervisor)?;
+
+        // If we are using the KVM hypervisor we meed to query for the new xsave2 size and update
+        // `XSAVE_FAM_LENGTH` accordingly.
+        #[cfg(feature = "kvm")]
+        {
+            // Ensure that XSAVE_STATE gets the correct size on the next call to get_state
+            let mut guard = XSAVE_FAM_LENGTH.write().unwrap();
+            Self::request_guest_amx_support()?;
+            // Obtain the number of bytes the kvm_xsave struct requires.
+            // This number is documented to always be at least 4096 bytes, but
+            let size = hypervisor.check_extension_int(kvm_ioctls::Cap::Xsave2);
+            // Reality check: We should at least have this number of bytes and probably more as we have enabled
+            // AMX tiles. If this is not the case, it is probably best to panic.
+            assert!(size >= 4096);
+            let fam_length = {
+                // Computation is documented in `[kvm_bindings::kvm_xsave2::len]`
+                ((size as usize) - size_of::<kvm_bindings::kvm_xsave>())
+                    .div_ceil(size_of::<kvm_bindings::__u32>())
+            };
+            *guard = fam_length;
+        }
+
+        #[cfg(not(feature = "kvm"))]
+        {
+            Self::request_guest_amx_support()?;
+        }
+
+        *amx_enabled_guard = true;
+        Ok(())
+    }
+
+    /// Checks whether the host supports AMX.
+    ///
+    /// The `hypervisor` is used to inform us about the
+    /// CPU vendor (AMX is currently only available on Intel CPUs).
+    fn amx_supported(hypervisor: &dyn Hypervisor) -> Result<(), AmxGuestSupportError> {
+        if !matches!(hypervisor.get_cpu_vendor(), CpuVendor::Intel) {
+            return Err(AmxGuestSupportError {
+                msg: "AMX is only available on Intel CPUs",
```

Comment:
I don't think this string approach is a good choice here. Please create a error enum with all the variants that describe possible failures - what you are doing here doesn't align with the rest of the code base unfortunately.

Reply:
Done

Reply:
Do you know git reflog? It helps to recover from messed up rebases - it's amazing :)